src/login/oidc/refresh/TokenRefresher.ts

```diff
@@ -0,0 +1,120 @@
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import { injectable, inject } from "tsyringe";
+import { IFetcher } from "../../../util/Fetcher";
+import { IStorageUtility } from "../../../localStorage/StorageUtility";
+import { IIssuerConfigFetcher } from "../IssuerConfigFetcher";
+import { IDpopHeaderCreator } from "../../../dpop/DpopHeaderCreator";
+import URL from "url-parse";
+import formurlencoded from "form-urlencoded";
+import { ITokenSaver } from "../redirectHandler/TokenSaver";
+
+export interface ITokenRefresher {
+  refresh(localUserId: string): Promise<void>;
+}
+
+@injectable()
+export default class TokenRefresher implements ITokenRefresher {
+  constructor(
+    @inject("fetcher") private fetcher: IFetcher,
+    @inject("issuerConfigFetcher")
+    private issuerConfigFetcher: IIssuerConfigFetcher,
+    @inject("storageUtility") private storageUtility: IStorageUtility,
+    @inject("dpopHeaderCreator") private dpopHeaderCreator: IDpopHeaderCreator
+  ) {}
+
+  async refresh(localUserId: string): Promise<void> {
+    // Get the refresh token and the issuer
+    const [refreshToken, issuer, clientId, clientSecret] = await Promise.all([
+      this.storageUtility.getForUser(localUserId, "refreshToken", true),
+      this.storageUtility.getForUser(localUserId, "issuer", true),
+      this.storageUtility.getForUser(localUserId, "clientId", true),
+      this.storageUtility.getForUser(localUserId, "clientSecret")
+    ]);
+
+    // Get the issuer config to find the token url
+    const issuerConfig = await this.issuerConfigFetcher.fetchConfig(
+      new URL(issuer as string)
+    );
+    if (
+      !issuerConfig.grantTypesSupported ||
+      !issuerConfig.grantTypesSupported.includes("refresh_token")
+    ) {
+      throw new Error(
+        `The issuer ${issuer} does not support the refresh token grant`
+      );
+    }
+    if (!issuerConfig.tokenEndpoint) {
+      throw new Error(`This issuer ${issuer} does not have a token endpoint`);
+    }
+
+    // Perform refresh token flow
+    const tokenRequestInit: RequestInit = {
+      method: "POST",
+      headers: {
+        DPoP: await this.dpopHeaderCreator.createHeaderToken(
+          issuerConfig.tokenEndpoint,
+          "POST"
+        ),
+        "content-type": "application/x-www-form-urlencoded"
+      },
```

Comment:
To avoid the cast on line 90, you might want to consider not to define this type so much narrower than it is. If you want to ensure it conforms to `RequestInit` still, you could do something like this:

```suggestion
    const tokenRequestInit: RequestInit & { headers: Record<string, string> } = {
      method: "POST",
      headers: {
        DPoP: await this.dpopHeaderCreator.createHeaderToken(
          issuerConfig.tokenEndpoint,
          "POST"
        ),
        "content-type": "application/x-www-form-urlencoded"
      },
```

Reply:
Cool. Done